File: flink-jobs/java/src/main/java/com/lakehouse/flink/KafkaToIcebergJob.java
package com.lakehouse.flink;

import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
import org.apache.flink.table.api.EnvironmentSettings;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Kafka to Iceberg Streaming Job (Java/Flink).
 *
 * <p>This job demonstrates a production-quality streaming pipeline that:
 * <ul>
 *   <li>Reads JSON events from Kafka topics</li>
 *   <li>Processes and transforms streaming data</li>
 *   <li>Writes to Iceberg tables with ACID guarantees</li>
 *   <li>Supports exactly-once processing semantics</li>
 * </ul>
 *
 * <p>Architecture:
 * <pre>
 * [Kafka Topic] -&gt; [Flink Streaming] -&gt; [Iceberg Table] -&gt; [MinIO Storage]
 *                         |
 *                  [PostgreSQL Catalog]
 * </pre>
 *
 * <p>Configuration is loaded from environment variables. See {@link JobConfig} for details.
 *
 * @author Streaming Lakehouse Lab
 * @version 0.1.0
 */
public class KafkaToIcebergJob {

    private static final Logger LOG = LoggerFactory.getLogger(KafkaToIcebergJob.class);

    /**
     * Main entry point for the Kafka to Iceberg streaming job.
     *
     * @param args Command-line arguments (not used)
     * @throws Exception If job execution fails
     */
    public static void main(String[] args) throws Exception {
        LOG.info("Starting Kafka to Iceberg streaming job");

        // Load configuration from environment
        JobConfig config = JobConfig.fromEnvironment();
        LOG.info("Loaded configuration: {}", config);

        // Create Flink execution environment
        StreamExecutionEnvironment env = createStreamEnvironment(config);

        // Create Table API environment
        StreamTableEnvironment tableEnv = createTableEnvironment(env);

        // Register Iceberg catalog
        registerIcebergCatalog(tableEnv, config);

        // Create source and sink tables
        createKafkaSourceTable(tableEnv, config);
        createIcebergSinkTable(tableEnv, config);

        // Execute streaming pipeline
        executeStreamingPipeline(tableEnv);

        LOG.info("Kafka to Iceberg job submitted successfully");
    }

    /**
     * Creates and configures the Flink streaming execution environment.
     *
     * <p>Configuration includes:
     * <ul>
     *   <li>Parallelism settings</li>
     *   <li>Checkpointing for fault tolerance</li>
     *   <li>State backend configuration</li>
     * </ul>
     *
     * @param config Job configuration
     * @return Configured StreamExecutionEnvironment
     */
    private static StreamExecutionEnvironment createStreamEnvironment(JobConfig config) {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        // Set parallelism
        env.setParallelism(config.getParallelism());

        // Enable checkpointing for exactly-once semantics
        env.enableCheckpointing(config.getCheckpointInterval());

        LOG.info("Created streaming environment with parallelism={}, checkpoint_interval={}ms",
                config.getParallelism(), config.getCheckpointInterval());

        return env;
    }

    /**
     * Creates the Flink Table API environment for SQL-based processing.
     *
     * @param env Streaming execution environment
     * @return Configured StreamTableEnvironment
     */
    private static StreamTableEnvironment createTableEnvironment(StreamExecutionEnvironment env) {
        EnvironmentSettings settings = EnvironmentSettings.inStreamingMode();
        StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env, settings);

        LOG.info("Created Table API environment");
        return tableEnv;
    }

    /**
     * Registers the Iceberg catalog with Flink Table API.
     *
     * <p>The catalog configuration includes:
     * <ul>
     *   <li>JDBC connection to PostgreSQL for metadata</li>
     *   <li>S3/MinIO connection for data files</li>
     *   <li>Warehouse location and namespace</li>
     * </ul>
     *
     * @param tableEnv Table environment
     * @param config Job configuration
     */
    private static void registerIcebergCatalog(StreamTableEnvironment tableEnv, JobConfig config) {
        String catalogName = config.getIcebergCatalogName();

        // Create Iceberg catalog
        String catalogDDL = String.format(
                "CREATE CATALOG IF NOT EXISTS %s WITH (\n" +
                        "  'type' = 'iceberg',\n" +
                        "  'catalog-type' = 'jdbc',\n" +
                        "  'uri' = '%s',\n" +
                        "  'jdbc.user' = '%s',\n" +
                        "  'jdbc.password' = '%s',\n" +
                        "  'warehouse' = '%s',\n" +
                        "  'io-impl' = 'org.apache.iceberg.aws.s3.S3FileIO',\n" +
                        "  's3.endpoint' = '%s',\n" +
                        "  's3.path-style-access' = 'true',\n" +
                        "  's3.access-key-id' = '%s',\n" +
                        "  's3.secret-access-key' = '%s'\n" +
                        ")",
                catalogName,
                config.getCatalogJdbcUrl(),
                config.getCatalogJdbcUser(),
                config.getCatalogJdbcPassword(),
                config.getIcebergWarehouse(),
                config.getS3Endpoint(),
                config.getS3AccessKey(),
                config.getS3SecretKey()
        );

        tableEnv.executeSql(catalogDDL);

        // Use the catalog
        tableEnv.useCatalog(catalogName);

        // Create and use namespace/database
        String namespace = config.getIcebergNamespace();
        tableEnv.executeSql(String.format("CREATE DATABASE IF NOT EXISTS %s", namespace));
        tableEnv.useDatabase(namespace);

        LOG.info("Registered Iceberg catalog: {}, namespace: {}", catalogName, namespace);
    }

    /**
     * Creates the Kafka source table for reading streaming events.
     *
     * <p>Table schema includes:
     * <ul>
     *   <li>event_id: Unique event identifier</li>
     *   <li>event_type: Event category</li>
     *   <li>user_id: User identifier</li>
     *   <li>timestamp_ms: Event timestamp</li>
     *   <li>payload: JSON event payload</li>
     *   <li>event_time: Computed watermark timestamp</li>
     * </ul>
     *
     * @param tableEnv Table environment
     * @param config Job configuration
     */
    private static void createKafkaSourceTable(StreamTableEnvironment tableEnv, JobConfig config) {
        String sourceDDL = String.format(
                "CREATE TABLE IF NOT EXISTS kafka_events (\n" +
                        "  event_id STRING,\n" +
                        "  event_type STRING,\n" +
                        "  user_id STRING,\n" +
                        "  timestamp_ms BIGINT,\n" +
                        "  payload STRING,\n" +
                        "  event_time AS TO_TIMESTAMP_LTZ(timestamp_ms, 3),\n" +
                        "  WATERMARK FOR event_time AS event_time - INTERVAL '5' SECOND\n" +
                        ") WITH (\n" +
                        "  'connector' = 'kafka',\n" +
                        "  'topic' = '%s',\n" +
                        "  'properties.bootstrap.servers' = '%s',\n" +
                        "  'properties.group.id' = 'flink-consumer',\n" +
                        "  'scan.startup.mode' = 'earliest-offset',\n" +
                        "  'format' = 'json',\n" +
                        "  'json.fail-on-missing-field' = 'false',\n" +
                        "  'json.ignore-parse-errors' = 'true'\n" +
                        ")",
                config.getKafkaTopicEvents(),
                config.getKafkaBootstrapServers()
        );

        tableEnv.executeSql(sourceDDL);
        LOG.info("Created Kafka source table: kafka_events");
    }

    /**
     * Creates the Iceberg sink table for storing processed events.
     *
     * <p>Table features:
     * <ul>
     *   <li>ACID transactions</li>
     *   <li>Time travel capabilities</li>
     *   <li>Schema evolution support</li>
     *   <li>Partitioned by date for efficient queries</li>
     * </ul>
     *
     * @param tableEnv Table environment
     * @param config Job configuration
     */
    private static void createIcebergSinkTable(StreamTableEnvironment tableEnv, JobConfig config) {
        String sinkDDL =
                "CREATE TABLE IF NOT EXISTS events_iceberg (\n" +
                        "  event_id STRING,\n" +
                        "  event_type STRING,\n" +
                        "  user_id STRING,\n" +
                        "  event_time TIMESTAMP(3),\n" +
                        "  timestamp_ms BIGINT,\n" +
                        "  payload STRING,\n" +
                        "  processing_time TIMESTAMP(3),\n" +
                        "  date_partition STRING,\n" +
                        "  PRIMARY KEY (event_id) NOT ENFORCED\n" +
                        ") PARTITIONED BY (date_partition)\n" +
                        "WITH (\n" +
                        "  'format-version' = '2',\n" +
                        "  'write.format.default' = 'parquet',\n" +
                        "  'write.metadata.compression-codec' = 'gzip',\n" +
                        "  'write.parquet.compression-codec' = 'snappy'\n" +
                        ")";

        tableEnv.executeSql(sinkDDL);
        LOG.info("Created Iceberg sink table: events_iceberg");
    }

    /**
     * Executes the streaming pipeline that reads from Kafka and writes to Iceberg.
     *
     * <p>Pipeline transformations:
     * <ul>
     *   <li>Read events from Kafka source</li>
     *   <li>Add processing timestamp</li>
     *   <li>Compute date partition for efficient queries</li>
     *   <li>Write to Iceberg with exactly-once semantics</li>
     * </ul>
     *
     * @param tableEnv Table environment
     */
    private static void executeStreamingPipeline(StreamTableEnvironment tableEnv) {
        String insertSQL =
                "INSERT INTO events_iceberg\n" +
                        "SELECT\n" +
                        "  event_id,\n" +
                        "  event_type,\n" +
                        "  user_id,\n" +
                        "  event_time,\n" +
                        "  timestamp_ms,\n" +
                        "  payload,\n" +
                        "  CURRENT_TIMESTAMP AS processing_time,\n" +
                        "  DATE_FORMAT(event_time, 'yyyy-MM-dd') AS date_partition\n" +
                        "FROM kafka_events";

        tableEnv.executeSql(insertSQL);
        LOG.info("Streaming pipeline started successfully");
    }
}

// MISSING_VALIDATION: Add data quality checks before writing to Iceberg
// MISSING_TEST: Add unit and integration tests
// MISSING_DOC: Add Javadoc examples and usage patterns

File: flink-jobs/python/kafka_to_iceberg_job.py
"""
Kafka to Iceberg Streaming Job (PyFlink).

This job demonstrates a complete streaming pipeline:
1. Reads JSON events from Kafka topic
2. Parses and validates event data
3. Writes to Iceberg table with ACID guarantees
4. Supports exactly-once processing semantics

Architecture:
    [Kafka Topic] -> [Flink Streaming] -> [Iceberg Table] -> [MinIO Storage]
                            |
                     [PostgreSQL Catalog]

Usage:
    python kafka_to_iceberg_job.py

Environment Variables:
    KAFKA_BOOTSTRAP_SERVERS: Kafka broker address (default: kafka:9092)
    KAFKA_TOPIC_EVENTS: Input Kafka topic (default: streaming.events)
    ICEBERG_CATALOG_NAME: Iceberg catalog name (default: iceberg_catalog)
    ... (see common.py for full list)
"""

import logging

from common import (
    FlinkConfig,
    get_catalog_properties,
    get_config,
    get_kafka_properties,
    setup_logging,
)
from pyflink.datastream import StreamExecutionEnvironment
from pyflink.table import (
    EnvironmentSettings,
    StreamTableEnvironment,
)

logger = logging.getLogger(__name__)


# ============================================================
# Job Configuration
# ============================================================


def create_stream_env(config: FlinkConfig) -> StreamExecutionEnvironment:
    """
    Create and configure Flink streaming environment.

    Args:
        config: Flink configuration

    Returns:
        Configured StreamExecutionEnvironment instance
    """
    env = StreamExecutionEnvironment.get_execution_environment()

    # Set parallelism
    env.set_parallelism(config.parallelism)

    # Enable checkpointing for fault tolerance
    env.enable_checkpointing(config.checkpoint_interval)

    logger.info(
        "Created Flink environment with parallelism=%d, checkpoint_interval=%dms",
        config.parallelism,
        config.checkpoint_interval,
    )

    return env


def create_table_env(
    stream_env: StreamExecutionEnvironment,
) -> StreamTableEnvironment:
    """
    Create Flink Table API environment.

    Args:
        stream_env: Streaming execution environment

    Returns:
        Configured StreamTableEnvironment instance
    """
    settings = EnvironmentSettings.in_streaming_mode()
    table_env = StreamTableEnvironment.create(stream_env, settings)

    logger.info("Created Flink Table environment")
    return table_env


# ============================================================
# Catalog Setup
# ============================================================


def register_iceberg_catalog(
    table_env: StreamTableEnvironment,
    config: FlinkConfig,
) -> None:
    """
    Register Iceberg catalog with Flink Table API.

    This configures the connection to the Iceberg catalog stored
    in PostgreSQL, with data files stored in MinIO (S3).

    Args:
        table_env: Flink Table environment
        config: Flink configuration
    """
    catalog_name = config.iceberg_catalog_name
    catalog_props = get_catalog_properties(config)

    # Create catalog
    table_env.execute_sql(
        f"""
        CREATE CATALOG IF NOT EXISTS {catalog_name}
        WITH (
            {', '.join(f"'{k}' = '{v}'" for k, v in catalog_props.items())}
        )
        """,
    )

    # Use the catalog
    table_env.use_catalog(catalog_name)

    # Create namespace/database if not exists
    table_env.execute_sql(
        f"CREATE DATABASE IF NOT EXISTS {config.iceberg_namespace}",
    )
    table_env.use_database(config.iceberg_namespace)

    logger.info(
        "Registered Iceberg catalog: %s, namespace: %s",
        catalog_name,
        config.iceberg_namespace,
    )


# ============================================================
# Source Configuration (Kafka)
# ============================================================


def create_kafka_source_table(
    table_env: StreamTableEnvironment,
    config: FlinkConfig,
) -> None:
    """
    Create Kafka source table for reading streaming events.

    Table schema:
        - event_id: STRING (unique event identifier)
        - event_type: STRING (event type/category)
        - user_id: STRING (user identifier)
        - timestamp_ms: BIGINT (event timestamp in milliseconds)
        - payload: STRING (JSON payload)
        - event_time: TIMESTAMP(3) (computed event time)

    Args:
        table_env: Flink Table environment
        config: Flink configuration
    """
    kafka_props = get_kafka_properties(config)

    table_env.execute_sql(
        f"""
        CREATE TABLE IF NOT EXISTS kafka_events (
            event_id STRING,
            event_type STRING,
            user_id STRING,
            timestamp_ms BIGINT,
            payload STRING,
            event_time AS TO_TIMESTAMP_LTZ(timestamp_ms, 3),
            WATERMARK FOR event_time AS event_time - INTERVAL '5' SECOND
        ) WITH (
            'connector' = 'kafka',
            'topic' = '{config.kafka_topic_events}',
            'properties.bootstrap.servers' = '{kafka_props["bootstrap.servers"]}',
            'properties.group.id' = '{kafka_props["group.id"]}',
            'scan.startup.mode' = 'earliest-offset',
            'format' = 'json',
            'json.fail-on-missing-field' = 'false',
            'json.ignore-parse-errors' = 'true'
        )
        """,
    )

    logger.info("Created Kafka source table: kafka_events")


# ============================================================
# Sink Configuration (Iceberg)
# ============================================================


def create_iceberg_sink_table(
    table_env: StreamTableEnvironment,
    config: FlinkConfig,
) -> None:
    """
    Create Iceberg sink table for storing processed events.

    Table features:
        - ACID transactions
        - Time travel capabilities
        - Schema evolution
        - Partition pruning

    Args:
        table_env: Flink Table environment
        config: Flink configuration
    """
    table_env.execute_sql(
        """
        CREATE TABLE IF NOT EXISTS events_iceberg (
            event_id STRING,
            event_type STRING,
            user_id STRING,
            event_time TIMESTAMP(3),
            timestamp_ms BIGINT,
            payload STRING,
            processing_time TIMESTAMP(3),
            date_partition STRING,
            PRIMARY KEY (event_id) NOT ENFORCED
        ) PARTITIONED BY (date_partition)
        WITH (
            'format-version' = '2',
            'write.format.default' = 'parquet',
            'write.metadata.compression-codec' = 'gzip',
            'write.parquet.compression-codec' = 'snappy'
        )
        """,
    )

    logger.info("Created Iceberg sink table: events_iceberg")


# ============================================================
# Data Processing Pipeline
# ============================================================


def run_streaming_job(config: FlinkConfig) -> None:
    """
    Execute the Kafka to Iceberg streaming job.

    Pipeline steps:
        1. Read from Kafka source table
        2. Transform and enrich data
        3. Write to Iceberg sink table
        4. Commit with exactly-once semantics

    Args:
        config: Flink configuration
    """
    # Create execution environments
    stream_env = create_stream_env(config)
    table_env = create_table_env(stream_env)

    # Register Iceberg catalog
    register_iceberg_catalog(table_env, config)

    # Create source and sink tables
    create_kafka_source_table(table_env, config)
    create_iceberg_sink_table(table_env, config)

    # Define the streaming query
    # Read from Kafka, transform, and write to Iceberg
    table_env.execute_sql(
        """
        INSERT INTO events_iceberg
        SELECT
            event_id,
            event_type,
            user_id,
            event_time,
            timestamp_ms,
            payload,
            CURRENT_TIMESTAMP AS processing_time,
            DATE_FORMAT(event_time, 'yyyy-MM-dd') AS date_partition
        FROM kafka_events
        """,
    )

    logger.info("Streaming job started successfully")


# ============================================================
# Main Entry Point
# ============================================================


def main() -> None:
    """
    Main entry point for the Kafka to Iceberg streaming job.

    Workflow:
        1. Load configuration from environment
        2. Setup logging
        3. Run streaming job
        4. Handle errors and cleanup
    """
    # Load configuration
    config = get_config()

    # Setup logging
    setup_logging(config.log_level)

    logger.info("Starting Kafka to Iceberg streaming job")
    logger.info("Configuration: %s", config.model_dump())

    try:
        # Run the streaming job
        run_streaming_job(config)

        logger.info("Job submitted successfully")

    except Exception:
        logger.exception("Failed to run streaming job")
        raise


if __name__ == "__main__":
    main()


# MISSING_VALIDATION: Add input data validation before writing to Iceberg
# MISSING_TEST: Add integration tests with test containers
# MISSING_DOC: Add metrics and monitoring integration
